In [3]:
import os 

# The normal imports
import numpy as np
import pandas as pd
pd.set_option('display.max_colwidth', -1)

# Import the stats library
from scipy import stats

# These are the plottinglibraries we'll use:
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.mlab as mlab
import seaborn as sns

# Command for plots to appear in the iPython Notebook
%matplotlib inline

#ignore warnings
import warnings
warnings.filterwarnings("ignore")

#Starting with the dataset that Dr.Drew helped clean.->highschools saved to .cvs file from graduations rates (2) notebook
wd = os.getcwd() #get working directory
#highschools = pd.read_csv(wd+'\\data\\highschools.csv', low_memory=False)
highschools = pd.read_csv(wd+'/data/highschools.csv', low_memory=False)
highschools.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 478 entries, 0 to 477
Columns: 294 entries, Unnamed: 0 to Percent GLP
dtypes: bool(8), float64(264), int64(9), object(13)
memory usage: 1.0+ MB


In [6]:
highschools.columns

Index(['Unnamed: 0', 'unit_code', 'type_cd_txt', 'category_cd',
       'school_type_txt', 'calendar_only_txt', 'title1_type_flg',
       'clp_ind_flg', 'focus_clp_flg', 'summer_program_flg',
       ...
       'State_Prin_Exp_Pct_10+_Years', 'School Code', 'Subject',
       'Percent Level 1', 'Percent Level 2', 'Percent Level 3',
       'Percent Level 4', 'Percent Level 5', 'Percent CCR', 'Percent GLP'],
      dtype='object', length=294)

In [8]:
pd.get_option('display.max_columns')

20

In [64]:
pd.set_option('display.max_columns', None)

In [13]:
pd.reset_option('display.max_columns')

In [12]:
columnNames = highschools.columns

In [16]:
columnNames

Index(['Unnamed: 0', 'unit_code', 'type_cd_txt', 'category_cd',
       'school_type_txt', 'calendar_only_txt', 'title1_type_flg',
       'clp_ind_flg', 'focus_clp_flg', 'summer_program_flg',
       ...
       'State_Prin_Exp_Pct_10+_Years', 'School Code', 'Subject',
       'Percent Level 1', 'Percent Level 2', 'Percent Level 3',
       'Percent Level 4', 'Percent Level 5', 'Percent CCR', 'Percent GLP'],
      dtype='object', length=294)

In [18]:
type(columnNames)

pandas.indexes.base.Index

In [21]:
columnNames.values

array(['Unnamed: 0', 'unit_code', 'type_cd_txt', 'category_cd',
       'school_type_txt', 'calendar_only_txt', 'title1_type_flg',
       'clp_ind_flg', 'focus_clp_flg', 'summer_program_flg',
       'asm_no_spg_flg', 'no_data_spg_flg', 'esea_status', 'student_num',
       'lea_avg_student_num', 'st_avg_student_num', 'Grad_project_status',
       'stem_flg', 'Avg_Class_Size_00', 'Avg_Class_Size_01',
       'Avg_Class_Size_02', 'Avg_Class_Size_03', 'Avg_Class_Size_04',
       'Avg_Class_Size_05', 'Avg_Class_Size_06', 'Avg_Class_Size_07',
       'Avg_Class_Size_08', 'Avg_Class_Size_Biology',
       'Avg_Class_Size_EnglishII', 'Avg_Class_Size_MathI',
       'Avg_Dist_Size_00', 'Avg_Dist_Size_01', 'Avg_Dist_Size_02',
       'Avg_Dist_Size_03', 'Avg_Dist_Size_04', 'Avg_Dist_Size_05',
       'Avg_Dist_Size_06', 'Avg_Dist_Size_07', 'Avg_Dist_Size_08',
       'Avg_Dist_Size_Biology', 'Avg_Dist_Size_EnglishII',
       'Avg_Dist_Size_MathI', 'Avg_State_Size_00', 'Avg_State_Size_01',
       'Avg_St

- Collect all column names starting with lea_ and st_

In [43]:
leaColNames = [col for col in columnNames if col.startswith('lea_')]
stColNames = [col for col in columnNames if col.startswith('st_')]
avgStateColNames = [col for col in columnNames if col.startswith('Avg_State_')]
avgClassColNames = [col for col in columnNames if col.startswith('Avg_Class_')]
avgDistColNames = [col for col in columnNames if col.startswith('Avg_Dist_')]

In [44]:
excludeColNames = leaColNames + stColNames + avgStateColNames + avgClassColNames + avgDistColNames

In [45]:
useColNames = [col for col in columnNames if col not in excludeColNames]

In [46]:
useColNames

['Unnamed: 0',
 'unit_code',
 'type_cd_txt',
 'category_cd',
 'school_type_txt',
 'calendar_only_txt',
 'title1_type_flg',
 'clp_ind_flg',
 'focus_clp_flg',
 'summer_program_flg',
 'asm_no_spg_flg',
 'no_data_spg_flg',
 'esea_status',
 'student_num',
 'Grad_project_status',
 'stem_flg',
 'total_expense_num',
 'salary_expense_pct',
 'benefits_expense_pct',
 'services_expense_pct',
 'supplies_expense_pct',
 'instruct_equip_exp_pct',
 'other_expense_pct',
 'federal_perpupil_num',
 'local_perpupil_num',
 'state_perpupil_num',
 'SPG Grade',
 'SPG Score',
 'Reading SPG Grade',
 'Reading  SPG Score',
 'Math SPG Grade',
 'Math SPG Score',
 'EVAAS Growth Status',
 'EVAAS Growth Score',
 'Overall Achievement Score',
 'Read Score',
 'Math Score',
 'Science Score',
 'Math I Score',
 'English II Score',
 'Biology Score',
 'The ACT Score',
 'ACT WorkKeys Score',
 'Math Course Rigor Score',
 'Cohort Graduation Rate Standard Score',
 'State Gap Compared',
 'GCE_ALL_School_Pct',
 'GCE_LEP_School_Pct',


In [47]:
len(useColNames)

151

> Count the number of missing values in 

## Digress to understand loc, iloc and ix
loc, iloc and ix are dataframe methods to access rows and columns. They are very flexible and powerful. 


In [51]:
highschools[useColNames].clp_ind_flg.sum()

56

In [52]:
highschools[useColNames].stem_flg.sum()

6

In [55]:
highschools[useColNames].esea_status.head()

0    Non_Esea
1    Non_Esea
2    Non_Esea
3    Non_Esea
4    Non_Esea
Name: esea_status, dtype: object

In [56]:
highschools[useColNames].esea_status.value_counts()

Non_Esea     459
Esea_Pass    17 
Esea_Fail    2  
Name: esea_status, dtype: int64

In [57]:
highschoolsFiltered = highschools[useColNames]

In [60]:
highschoolsFiltered.loc[highschoolsFiltered.esea_status == "Esea_Pass", ['unit_code', 'school_type_txt', 'esea_status']]

unit_code school_type_txt esea_status
10   40306     Regular School  Esea_Pass 
72   181318    Regular School  Esea_Pass 
113  260455    Regular School  Esea_Pass 
139  320314    Regular School  Esea_Pass 
141  320325    Regular School  Esea_Pass 
185  400308    Regular School  Esea_Pass 
210  420358    Regular School  Esea_Pass 
283  600386    Regular School  Esea_Pass 
296  600576    Regular School  Esea_Pass 
297  600579    Regular School  Esea_Pass 
346  740344    Regular School  Esea_Pass 
348  740374    Regular School  Esea_Pass 
349  740388    Regular School  Esea_Pass 
364  780325    Regular School  Esea_Pass 
365  780342    Regular School  Esea_Pass 
450  930352    Regular School  Esea_Pass 
458  960335    Regular School  Esea_Pass

In [68]:
pd.get_option('display.max_rows')

60

In [71]:
pd.set_option('display.max_rows', None)

In [70]:
pd.reset_option('display.max_rows')

## Find missing values

Use the entire highschool dataframe and find the columns with highest number of missing values.

In [61]:
highschools.shape

(478, 294)

Out of this 294 columns, how many of them have 100% missing data ? 

In [65]:
highschools.isnull().tail()

Unnamed: 0 unit_code type_cd_txt category_cd school_type_txt  \
473  False      False     False       False       False            
474  False      False     False       False       False            
475  False      False     False       False       False            
476  False      False     False       False       False            
477  False      False     False       False       False            

    calendar_only_txt title1_type_flg clp_ind_flg focus_clp_flg  \
473  False             False           False       False          
474  False             False           False       False          
475  False             False           False       False          
476  False             False           False       False          
477  False             False           False       False          

    summer_program_flg asm_no_spg_flg no_data_spg_flg esea_status student_num  \
473  False              False          False           False       False        
474  False              False          False           False       False        
475  False              False          False           False       False        
476  False              False          False           False       False        
477  False              False          False           False       False        

    lea_avg_student_num st_avg_student_num Grad_project_status stem_flg  \
473  False               False              False               False     
474  False               False              False               False     
475  False               False              False               False     
476  False               False              False               False     
477  False               False              False               False     

    Avg_Class_Size_00 Avg_Class_Size_01 Avg_Class_Size_02 Avg_Class_Size_03  \
473  False             False             False             False              
474  True              True              True              True               
475  True              True              True              True               
476  True              True              True              True               
477  True              True              True              True               

    Avg_Class_Size_04 Avg_Class_Size_05 Avg_Class_Size_06 Avg_Class_Size_07  \
473  False             False             False             True               
474  True              True              True              True               
475  True              True              True              True               
476  True              True              True              True               
477  True              True              True              True               

    Avg_Class_Size_08 Avg_Class_Size_Biology Avg_Class_Size_EnglishII  \
473  True              True                   True                      
474  True              False                  False                     
475  True              False                  False                     
476  True              False                  False                     
477  True              False                  False                     

    Avg_Class_Size_MathI Avg_Dist_Size_00 Avg_Dist_Size_01 Avg_Dist_Size_02  \
473  True                 False            False            False             
474  False                True             True             True              
475  False                True             True             True              
476  False                True             True             True              
477  False                True             True             True              

    Avg_Dist_Size_03 Avg_Dist_Size_04 Avg_Dist_Size_05 Avg_Dist_Size_06  \
473  False            False            False            False             
474  True             True             True             True              
475  True             True             True             True              
476  True             True             True             True              
477  True      

In [66]:
highschools.tail()

Unnamed: 0 unit_code type_cd_txt category_cd school_type_txt  \
473  2398        990304    Public      A           Regular School   
474  2404        990322    Public      H           Regular School   
475  2407        990326    Public      H           Regular School   
476  2409        990332    Public      H           Regular School   
477  2419        995330    Public      H           Regular School   

        calendar_only_txt title1_type_flg clp_ind_flg focus_clp_flg  \
473  Traditional Calendar  True            False       False          
474  Traditional Calendar  False           False       False          
475  Traditional Calendar  False           False       False          
476  Traditional Calendar  False           False       False          
477  Traditional Calendar  False           False       False          

    summer_program_flg asm_no_spg_flg no_data_spg_flg esea_status  \
473  True               False          False           Non_Esea     
474  False              False          False           Non_Esea     
475  False              False          False           Non_Esea     
476  False              False          False           Non_Esea     
477  True               False          False           Non_Esea     

     student_num  lea_avg_student_num  st_avg_student_num Grad_project_status  \
473  332.0        188.0                511                 False                
474  893.0        571.0                860                 False                
475  609.0        571.0                860                 False                
476  212.0        571.0                860                 False                
477  691.0        691.0                860                 False                

    stem_flg  Avg_Class_Size_00  Avg_Class_Size_01  Avg_Class_Size_02  \
473  False    18.0               19.0               17.0                
474  False   NaN                NaN                NaN                  
475  False   NaN                NaN                NaN                  
476  False   NaN                NaN                NaN                  
477  False   NaN                NaN                NaN                  

     Avg_Class_Size_03  Avg_Class_Size_04  Avg_Class_Size_05  \
473  20.0               19.0               25.0                
474 NaN                NaN                NaN                  
475 NaN                NaN                NaN                  
476 NaN                NaN                NaN                  
477 NaN                NaN                NaN                  

     Avg_Class_Size_06  Avg_Class_Size_07  Avg_Class_Size_08  \
473  17.0              NaN                NaN                  
474 NaN                NaN                NaN                  
475 NaN                NaN                NaN                  
476 NaN                NaN                NaN                  
477 NaN                NaN                NaN                  

     Avg_Class_Size_Biology  Avg_Class_Size_EnglishII  Avg_Class_Size_MathI  \
473 NaN                     NaN                       NaN                     
474  19.0                    22.0                      16.0                   
475  21.0                    19.0                      20.0                   
476  24.0                    23.0                      21.0                   
477  17.0                    16.0                      19.0                   

     Avg_Dist_Size_00  Avg_Dist_Size_01  Avg_Dist_Size_02  Avg_Dist_Size_03  \
473  18.0              17.0              16.0              16.0               
474 NaN               NaN               NaN               NaN                 
475 NaN               NaN               NaN               NaN                 
476 NaN               NaN               NaN               NaN                 
477 NaN               NaN               NaN               NaN                 

     Avg_Dist_Size_04  Avg_Dist_Size_05  Avg_Dist_Size_06  Avg_Dist_Size_07  \
473  19.0              

Use `isnull()` to figure out how many missing values are present per column.

In [72]:
highschools.isnull().sum()

Unnamed: 0                                        0  
unit_code                                         0  
type_cd_txt                                       0  
category_cd                                       0  
school_type_txt                                   0  
calendar_only_txt                                 0  
title1_type_flg                                   0  
clp_ind_flg                                       0  
focus_clp_flg                                     0  
summer_program_flg                                0  
asm_no_spg_flg                                    0  
no_data_spg_flg                                   0  
esea_status                                       0  
student_num                                       0  
lea_avg_student_num                               0  
st_avg_student_num                                0  
Grad_project_status                               0  
stem_flg                                          0  
Avg_Class_Size_00           

In [73]:
type(highschools.isnull().sum())

pandas.core.series.Series

Get the column names whose value is equal to 478. 

In [75]:
(highschools.isnull().sum() == 478).sum()

47

First let's store the result of `highschools.isnull().sum()` which is a Series.

In [76]:
missingValues = highschools.isnull().sum()

In [81]:
len(list(missingValues[missingValues == 478].index))

47

In [83]:
allMissingValues = list(missingValues[missingValues == 478].index)

Filter out these columns from the dataframe `highschools`.

In [84]:
highschools.drop(allMissingValues, axis=1, inplace=True)

In [85]:
highschools.shape

(478, 247)

In [86]:
247 + 47

294

Next, we will drop columns which have 95% of missing data.

In [87]:
highschools.isnull().sum()

Unnamed: 0                                        0  
unit_code                                         0  
type_cd_txt                                       0  
category_cd                                       0  
school_type_txt                                   0  
calendar_only_txt                                 0  
title1_type_flg                                   0  
clp_ind_flg                                       0  
focus_clp_flg                                     0  
summer_program_flg                                0  
asm_no_spg_flg                                    0  
no_data_spg_flg                                   0  
esea_status                                       0  
student_num                                       0  
lea_avg_student_num                               0  
st_avg_student_num                                0  
Grad_project_status                               0  
stem_flg                                          0  
Avg_Class_Size_00           

In [88]:
478 * 0.95

454.09999999999997

Drop columns which have more than 454 or more missing values

In [89]:
missingValues = highschools.isnull().sum()

In [94]:
missing95Percent = list(missingValues[missingValues >= 478 * 0.95].index)

In [95]:
missing95Percent

['Avg_Class_Size_00',
 'Avg_Class_Size_01',
 'Avg_Class_Size_02',
 'Avg_Class_Size_03',
 'Avg_Class_Size_04',
 'Avg_Class_Size_05',
 'Avg_Class_Size_06',
 'Avg_Class_Size_07',
 'Avg_Class_Size_08',
 'Avg_Dist_Size_00',
 'Avg_Dist_Size_01',
 'Avg_Dist_Size_02',
 'Avg_Dist_Size_03',
 'Avg_Dist_Size_04',
 'Avg_Dist_Size_05',
 'Avg_Dist_Size_06',
 'Avg_Dist_Size_07',
 'Avg_Dist_Size_08',
 'Avg_State_Size_00',
 'Avg_State_Size_01',
 'Avg_State_Size_02',
 'Avg_State_Size_03',
 'Avg_State_Size_04',
 'Avg_State_Size_05',
 'Avg_State_Size_06',
 'Avg_State_Size_07',
 'Avg_State_Size_08',
 'Reading SPG Grade',
 'Reading  SPG Score',
 'Math SPG Grade',
 'Math SPG Score',
 'Read Score',
 'Math Score',
 'Science Score',
 'GCE_ALL_School_Pct',
 'GCE_SWD_School_Pct',
 'PASSED_EOG_School_Pct',
 'PROMOTED_School_Pct',
 'GCE_ALL_Dist_Pct',
 'GCE_RPF_Dist_Pct',
 'GCE_SWD_Dist_Pct',
 'PASSED_EOG_Dist_Pct',
 'PASSED_LAA_Dist_Pct',
 'PROMOTED_Dist_Pct',
 'RETAINED_Dist_Pct',
 'GCE_ALL_State_Pct',
 'GCE_SWD_S

In [96]:
len(missing95Percent)

52

Drop these columns from the `highschools` dataframe.

In [97]:
highschools.drop(missing95Percent, axis=1, inplace=True)

In [98]:
highschools.shape

(478, 195)

In [99]:
highschools.isnull().sum()

Unnamed: 0                                        0  
unit_code                                         0  
type_cd_txt                                       0  
category_cd                                       0  
school_type_txt                                   0  
calendar_only_txt                                 0  
title1_type_flg                                   0  
clp_ind_flg                                       0  
focus_clp_flg                                     0  
summer_program_flg                                0  
asm_no_spg_flg                                    0  
no_data_spg_flg                                   0  
esea_status                                       0  
student_num                                       0  
lea_avg_student_num                               0  
st_avg_student_num                                0  
Grad_project_status                               0  
stem_flg                                          0  
Avg_Class_Size_Biology      

OK, so now we have `highschools` dataset with 195 columns and 478 rows. Our next task is to find out how many of them are categorical variables among these. So get the list of categorical variables from this dataframe.

In [101]:
highschools.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 478 entries, 0 to 477
Data columns (total 195 columns):
Unnamed: 0                                        int64
unit_code                                         object
type_cd_txt                                       object
category_cd                                       object
school_type_txt                                   object
calendar_only_txt                                 object
title1_type_flg                                   bool
clp_ind_flg                                       bool
focus_clp_flg                                     bool
summer_program_flg                                bool
asm_no_spg_flg                                    bool
no_data_spg_flg                                   bool
esea_status                                       object
student_num                                       float64
lea_avg_student_num                               float64
st_avg_student_num                                

In [116]:
boolHighSchool = highschools.loc[:, highschools.dtypes == bool]

In [119]:
type(boolHighSchool)

pandas.core.frame.DataFrame

In [120]:
boolColumns = list(boolHighSchool.columns)

In [124]:
highschools[boolColumns].apply(sum)

title1_type_flg        38 
clp_ind_flg            56 
focus_clp_flg          53 
summer_program_flg     179
asm_no_spg_flg         0  
no_data_spg_flg        0  
Grad_project_status    177
stem_flg               6  
dtype: int64

**Next task is to combine the description with these stats**

In [136]:
# This is a simple function to pull column description
DataDict = pd.read_csv(wd+'/data/dictionary.csv', encoding = "ISO-8859-1")
DataDict.head()

COLUMN_NAME                       DESCRIPTION
0  year         School Year                     
1  unit_code    Code to identifySchool/LEA/State
2  school_name  School Name                     
3  lea_name     LEA Name                        
4  state_name   State Name

In [137]:
DataDict.index

RangeIndex(start=0, stop=389, step=1)

In [138]:
DataDict.columns

Index(['COLUMN_NAME', 'DESCRIPTION'], dtype='object')

In [139]:
#DataDict.set_index('COLUMN_NAME', inplace=True)
DataDict.head(20)

COLUMN_NAME  \
0   year                
1   unit_code           
2   school_name         
3   lea_name            
4   state_name          
5   category_cd         
6   street_ad           
7   scity_ad            
8   state_ad            
9   szip_ad             
10  vphone_ad           
11  type_cd             
12  closed_ind          
13  new_ind             
14  super_nm            
15  url_ad              
16  grade_range_cd      
17  calendar_type_txt   
18  school_type_txt     
19  calendar_only_txt   

                                                                                            DESCRIPTION  
0   School Year                                                                                          
1   Code to identifySchool/LEA/State                                                                     
2   School Name                                                                                          
3   LEA Name                                                                                             
4   State Name                                                                                           
5   Category code E, M, H, I, T, or A                                                                    
6   Street address                                                                                       
7   City name                                                                                            
8   State                                                                                                
9   ZIP code                                                                                             
10  Phone Number                                                                                         
11  Charter or public school                                                                             
12  1=School is closed, 0=School is currently open                                                       
13  1=School is new this year, 0=School is not new\n*If a school is new this year, no data is available  
14  Superintendent’s Name                                                                                
15  URL Address of the school or district                                                                
16  Range of grades offered                                                                              
17  Description of school calendar and school type                                                       
18  Description of school type                                                                           
19  Description of Calendar

In [143]:
type(boolColumns)
boolColumns

['title1_type_flg',
 'clp_ind_flg',
 'focus_clp_flg',
 'summer_program_flg',
 'asm_no_spg_flg',
 'no_data_spg_flg',
 'Grad_project_status',
 'stem_flg']

In [148]:
DataDict[DataDict.COLUMN_NAME == 'lea_name']

COLUMN_NAME DESCRIPTION
3    lea_name    LEA Name  
37   lea_name    LEA Name  
46   lea_name    LEA Name  
84   lea_name    LEA Name  
99   lea_name    LEA Name  
130  lea_name    LEA Name  
137  lea_name    LEA Name  
187  lea_name    LEA Name  
211  lea_name    LEA Name  
266  lea_name    LEA Name  
277  lea_name    LEA Name

In [147]:
DataDict

COLUMN_NAME  \
0    year                             
1    unit_code                        
2    school_name                      
3    lea_name                         
4    state_name                       
5    category_cd                      
6    street_ad                        
7    scity_ad                         
8    state_ad                         
9    szip_ad                          
10   vphone_ad                        
11   type_cd                          
12   closed_ind                       
13   new_ind                          
14   super_nm                         
15   url_ad                           
16   grade_range_cd                   
17   calendar_type_txt                
18   school_type_txt                  
19   calendar_only_txt                
20   title1_type_cd                   
21   esea_status                      
22   grad_project_status              
23   student_num                      
24   lea_avg_student_num              
25   st_avg_student_num               
26   welcome_ltr_url                  
27   stem                             
28   county                           
29   clp_ind                          
30   focus_clp_ind                    
31   summer_program_ind               
32   asm_no_spg_ind                   
33   no_data_spg_ind                  
34   year                             
35   unit_code                        
36   school_name                      
37   lea_name                         
38   state_name                       
39   category_cd                      
40   level                            
41   size                             
42   lea_size                         
43   st_size                          
44   unit_code                        
45   school_name                      
46   lea_name                         
47   state_name                       
48   category_cd                      
49   total_expense_num                
50   lea_ total_expense_num           
51   st_ total_expense_num            
52   salary_expense_pct               
53   lea_salary_expense_pct           
54   st_ salary_expense_pct           
55   benefits_expense_pct             
56   lea_ benefits_expense_pct        
57   st_ benefits_expense_pct         
58   services_expense_pct             
59   lea_ services_expense_pct        
60   st_ services_expense_pct         
61   supplies_expense_pct             
62   lea_supplies_expense_pct         
63   st_supplies_expense_pct          
64   instruct_equip_exp_pct           
65   lea_ instruct_equip_exp_pct      
66   st_ instruct_equip_exp_pct       
67   other_expense_pct                
68   lea_ other_expense_pct           
69   st_ other_expense_pct            
70   lea_ building_expense_pct        
71   st_ building_expense_pct         
72   federal_perpupil_num             
73   lea_ federal_perpupil_num        
74   st_ federal_perpupil_num         
75   local_perpupil_num               
76   lea_ local_perpupil_num          
77   st_ local_perpupil_num           
78   state_perpupil_num               
79   lea_ state_perpupil_num          
80   st_ state_perpupil_num           
81   year                             
82   unit_code                        
83   school_name                      
84   lea_name                         
85   state_name                       
86   category_cd                      
87   passed_bog_and_or_eog            
88   exemption                        
89   outcome                          
90   num_students                     
91   lea_num_students                 
92   st_num_students                  
93   pct_students                     
94   lea_pct_students                 
95   st_pct_students                  
96   year                             
97   unit_code                        
98   school_name                      
99   lea_name                         
100  state_name                       
101  category_cd                      
102  sa

In [131]:
DataDict.shape

(389, 1)

In [135]:
DataDict.columns

Index(['DESCRIPTION'], dtype='object')

#### Business Understanding [10]

Describe the purpose of the data set you selected (i.e., why was this data collected in the first place?). Describe how you would define and measure the outcomes from the dataset. That is, why is this data important and how do you know if you have mined useful knowledge from the dataset? 
How would you measure the effectiveness of a good prediction algorithm? Be specific.

#### Data Meaning Type  [10]

Describe the meaning and type of data (scale, values, etc.) for each attribute in the data file.

#### Data Quality [15]

Verify data quality: Explain any missing values, duplicate data, and outliers. Are those mistakes? How do you deal with these problems? Give justifications for your methods.

In [4]:
#Get dimensions of the dataframe that we working with
highschools.shape

(478, 294)

In [5]:
#Make a copy of the dataset to work with
HighschoolData = highschools.copy()

##### *Missing Data Statistics*

In this section we explore missing data in the highschool data that was created from the original north carolina data. After exploring the missing data we will discusss the approach for handling them.

In [18]:
#Check for missing values
Temp = pd.DataFrame(HighschoolData.isnull().sum())
#print(Temp)
Temp.columns =['DataMissing']
#Columns with atleast 1 missing value
MissingCount = Temp[Temp.DataMissing>0]

#sort 
MissingCount = MissingCount.sort_values('DataMissing',ascending=False)
#basic starts on missing data
print("##################################################")
print("###", '             Stats on missing data        ',  "###")
print("##################################################")
print('No of rows in HighschoolData: ',len(HighschoolData))
print('No of Columns in HighschoolData: ',len(HighschoolData.columns))
print('No. of Columns with Missing data: ',len(MissingCount))
print('No  of Complete data columns: ', len(HighschoolData.columns) - len(MissingCount))
print('No. of Columns with 100% missing Values: ',sum(1 for item in MissingCount.DataMissing if item==len(HighschoolData)))
print('No. of Columns more than 95% missing Values: ',sum(1 for item in MissingCount.DataMissing if item>=0.95*len(HighschoolData)))
print('% of Columns with some Missing data: ',round(float(100*len(MissingCount))/len(HighschoolData.columns),3),'%')
print('% of Columns with 100% Missing data: ',round(float(100*sum(1 for item in MissingCount.DataMissing if item==len(HighschoolData)))/len(HighschoolData.columns),3),'%')
print('% of Columns with more than 95% Missing data: ',round(float(100*sum(1 for item in MissingCount.DataMissing if item>=0.95*len(HighschoolData)))/len(HighschoolData.columns),3),'%')

print(" ")
print("##################################################")
print("###",'       Columns with >95% missing data     ',    "###")
print("##################################################")
#add a columnName for Bar charts plot
MissingCount['ColumnName'] = MissingCount.index
MissingCount['Missing %'] = 100*(MissingCount.DataMissing/len(HighschoolData))
MissingCount = MissingCount.reset_index()
del MissingCount['index']
print((MissingCount.head(100)))

##################################################
###              Stats on missing data         ###
##################################################
No of rows in HighschoolData:  478
No of Columns in HighschoolData:  294
No. of Columns with Missing data:  168
No  of Complete data columns:  126
No. of Columns with 100% missing Values:  47
No. of Columns more than 95% missing Values:  99
% of Columns with some Missing data:  57.143 %
% of Columns with 100% Missing data:  15.986 %
% of Columns with more than 95% Missing data:  33.673 %
 
##################################################
###        Columns with >95% missing data      ###
##################################################
    DataMissing                ColumnName   Missing %
0   478          Percent GLP               100.000000
1   478          total_expense_num         100.000000
2   478          GCE_RPF_State_Pct         100.000000
3   478          GCE_PRM_State_Pct         100.000000
4   478          GCE_LEP_State_

##### Approach for handling missing data to be discussed with the rest of the team
From above we note, of the 293 columns in HighSchoolData, 168 (57.14%) of them have some missing data. 47 (15.99%) of columns have all the data missing where 99 (33.67%) of the columns miss more than 95% of the data. For columns with 100% missing data we are left but to wonder if this were new fields introduced recently and the schools are not fully informed to collect this data or the data for not just collected by mistake. On the other hand, for the partally completed datasets, some schools migh have left this columns blank if they didn't apply to the.

For the sake of this analysis we will drop the columns that have `>95` missing data rather than try complete the missing data with either mean, median or mode. If we complete this large amount of columns, with predetermined data, our model maybe 

In [20]:
#All columns from Original schoolData
AllColumns =HighschoolData.columns 

# All columns with some missing value
ColumnsWithMissingdata = MissingCount.ColumnName 

#columns with more than 95% missing data
Columns2Drop = MissingCount.ColumnName.head(39) 

#columns with missing values that are kept
Columns2Keep = MissingCount.ColumnName.tail(len(MissingCount)-39) 

#All the other columns except those with >95% missing data
SelectedColumns = list(set(AllColumns)-set(Columns2Drop)) 

#new dataset, columns with >95% missing data dropped

schoolDataNew = HighschoolData[SelectedColumns]

## Peek at New Dataframe
print(HighschoolData)
print(schoolDataNew.info())
schoolDataNew.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 478 entries, 0 to 477
Columns: 255 entries, total_specialized_courses to PASSED_EOG_State_Pct
dtypes: bool(8), float64(225), int64(9), object(13)
memory usage: 926.2+ KB
None


total_specialized_courses  Math Score Reading SPG Grade  lea_wap_num  \
0 NaN                        NaN          NaN               1380.0        
1 NaN                        NaN          NaN               1380.0        

   st_flicensed_teach_pct  st_not_highqual_class_all_pct  \
0  0.897                   0.037                           
1  0.897                   0.037                           

   PASSED_EOG_School_Pct  tchyrs_0thru3_pct type_cd_txt  PASSED_LAA_Dist_Pct  \
0 NaN                     0.000              Public     NaN                    
1 NaN                     0.186              Public     NaN                    

           ...           Avg_Class_Size_05  _1yr_tchr_trnovr_pct  \
0          ...          NaN                 0.000                  
1          ...          NaN                 0.128                  

   st_highqual_class_hp_pct  Grad_project_status  Math Course Rigor Score  \
0  0.964                     False               NaN                        
1  0.964                     False               NaN                        

   lea_long_susp_per_c_num  Avg_Class_Size_04  st_highqual_class_pct  \
0  0.23                    NaN                 0.964                   
1  0.23                    NaN                 0.964                   

   Avg_Class_Size_03  PASSED_EOG_State_Pct  
0 NaN                NaN                    
1 NaN                NaN                    

[2 rows x 255 columns]

#### Simple Statistics [10]
Visualize appropriate statistics (e.g., range, mode, mean, median, variance, counts) for a subset of attributes. Describe anything meaningful you found from this or if you found something potentially interesting. Note: You can also use data from other sources for comparison. Explain why the statistics run are meaningful. 

#### Visualize Attributes [15]
Visualize the most interesting attributes (at least 5 attributes, your opinion on what is interesting). Important: Interpret the implications for each visualization. Explain for each attribute why the chosen visualization is appropriate.

#### Explore Joint Attributes [15]
Visualize relationships between attributes: Look at the attributes via scatter plots, correlation, cross-tabulation, group-wise averages, etc. as appropriate. Explain any interesting relationships.

#### New Features [5]
Are there other features that could be added to the data or created from existing features? Which ones?

#### Exceptional Work [10]
You have free reign to provide additional analyses. One idea: implement dimensionality reduction, then visualize and interpret the results.

#### 1. Data Dictionary
Since this datasets has numerous columns we needed a fast way to quickly find ColumnName description for easy reference. For exceptional work, we created a function to quickly pull the data from csv datafile. This involved converting the pdf to excel and formating the data for easy import into pandas. The code below is a working code for our data dictionary.

In [15]:
# This is a simple function to pull column description
DataDict = pd.read_csv(wd+'\\data\\dictionary.csv', encoding = "ISO-8859-1")
DataDict.head()
#DataDict = DataDict.columns['COLUMN_NAME', 'DESCRIPTION']
def get_ColDescription(colname = 'Year'):
    colName = input("Enter column name to check description in Dictionary. You can enter multiple columns separated by comma: ")
    
    print('You entered: ', colName.strip())
    temp = pd.DataFrame()
    colNames = colName.split(',')
    
    try:
        for i in range(0,len(colNames)):
            get = (DataDict[DataDict.COLUMN_NAME==colNames[i].strip().lower()])
            temp = temp.append(get)
        return(temp)
    except Exception as e:
        print(e.args) 

get_ColDescription()

Enter column name to check description in Dictionary. You can enter multiple columns separated by comma: A, T, M, I, E
You entered:  A, T, M, I, E


COLUMN_NAME                                            DESCRIPTION
373  a           School has elementary, middle, and high school grades
372  t           School has middle and high school grades             
369  m           School has middle schools grades (6-8)               
371  i           School has elementary and middle school grades       
368  e           School has elementary schools grades (PK-5)